pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,229 @@
+                    'pubsub.producer.batch', len(new_messages))
+
+            await message_queue.join()
+
+    async def subscribe(subscription: str,  # pylint: disable=too-many-locals
+                        handler: ApplicationHandler,
+                        *,
+                        num_workers: int = 1,
+                        max_messages: int = 100,
+                        ack_window: float = 0.3,
+                        ack_deadline_cache_timeout: int = 10,
+                        consumer_pool_size: int = 1,
+                        subscriber_client: Optional[SubscriberClient] = None,
+                        metrics_client: Optional[MetricsAgent] = None
+                        ) -> None:
+        async with aiohttp.ClientSession() as session:
+            ack_queue: 'asyncio.Queue[str]' = asyncio.Queue(
+                maxsize=(max_messages * num_workers))
+            subscriber = subscriber_client or SubscriberClient(session=session)
+            ack_deadline_cache = AckDeadlineCache(subscriber,
+                                                  subscription,
+                                                  ack_deadline_cache_timeout)
+            tasks = []
+            try:
+                tasks.append(asyncio.ensure_future(
+                    acker(subscription, ack_queue, subscriber,
+                          ack_window=ack_window, metrics_client=metrics_client)
+                ))
+                for _ in range(num_workers):
+                    q: MessageQueue = asyncio.Queue(
+                        maxsize=max_messages)
+                    tasks.append(asyncio.ensure_future(
+                        consumer(q,
+                                 handler,
+                                 ack_queue,
+                                 ack_deadline_cache,
+                                 consumer_pool_size,
+                                 metrics_client=metrics_client)
+                    ))
+                    tasks.append(asyncio.ensure_future(
+                        producer(subscription,
+                                 q,
+                                 subscriber,
+                                 max_messages=max_messages,
+                                 metrics_client=metrics_client)
+                    ))
+
+                _done, pending = await asyncio.wait(
+                    tasks, return_when=asyncio.FIRST_COMPLETED)
+                log.info(
+                    'One of subsriber workers exited, stopping the rest...')
+                for task in pending:
+                    task.cancel()
+                await asyncio.wait(pending, return_when=asyncio.ALL_COMPLETED)
+            except Exception:
+                log.exception('Subscriber exited')
+                for task in tasks:
+                    task.cancel()
+                await asyncio.wait(tasks, return_when=asyncio.ALL_COMPLETED)
+            raise asyncio.CancelledError('One of subscription workers exited.')
```

Comment:
@jonathan-johnston we were discussing error handling here. My latest take: 
> I added exception handling now, so what should happen:
>1. If any subscriber code throws during tasks creation, we catch it, cancel already created tasks, wait for them to complete and raise CancelledError
>2. If any of the tasks complete, we consider this an unrecoverable crash, cancel all other tasks, wait for them to complete and raise CancelledError.

Reply:
re:exception sounds right to me

graceful shutdown is always nice. Is there any common way to gracefully shutdown asyncio tasks? Quick search didn't yield much results for me...

Reply:
I think `.cancel()` is the graceful way to do things! Or at least, I certainly haven't heard of any other options.

Reply:
I mean, can we somehow determine we're being cancelled and perform a cleanup? Otherwise we should be 100% compliant with cancel interface :D

Reply:
I mean, your shutdown code looks good here, I think it's more a matter of the potential plumbing for shutting down the overall subscriber. Eg. you could imagine something like the following (switched to OO to make the shared variables between methods a bit clearer, but of course could be done via eg. return values)
```python
class Subscriber:
    def __init__(self):
        self._tasks = []

    async def subscribe(self):
        # your existing code, minus the shutdown stuff at the end
        self._tasks = tasks
        try:
            await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
         except CancelledError:
             pass
         except Exception:
             log.exception('Subscriber exited')
         finally:
             self.close()

    def close(self):
        for task in self._tasks:  # Even better would be to shut down more gracefully, eg. producers before consumers
            task.cancel()  # TODO: this might need an `if not task.cancelled` to avoid double-cancelling? Is double-cancelling even a problem?
```

The major changes above being that calling the `close()` method and thus stopping our subscriber can gracefully shutdown tasks in the correct order and propogate the `CancelledError` to the `await subscribe()` call the user may have -- that is no longer treated as an error, but actual errors are still reported and cause "as graceful a shutdown as possible"

Reply:
The typical way to ensure a "graceful" shutdown is via `asyncio.Event` which is checked each loop in the tasks instead of `while True:`. That said, you should be careful with this method because it doesn't stop blocking actions (ex. `await` on a blocking queue) from making your program hang instead of shut down.

Reply:
Huh, yeah I see why graceful shutdown might be a thing of the future